File: Python/ByTier/E/1726_E_Almost_Perfect.py
# Problem: CF 1726 E - Almost Perfect
# https://codeforces.com/contest/1726/problem/E

"""
Almost Perfect Permutation Counting

This code computes the number of "almost perfect" permutations of length n.
A permutation p is almost perfect if for every index i, |p[i] - p^{-1}[i]| <= 1,
where p^{-1} is the inverse permutation.

The solution uses dynamic programming with memoization and generates the counts
using a recurrence relation. The recurrence relation is derived from the structure
of valid permutations where elements can only be at most 1 position away from
their original index.

Time Complexity: O(n) preprocessing + O(n) per query
Space Complexity: O(n) for precomputed values

Algorithms/Techniques:
- Dynamic Programming with recurrence relation
- Modular arithmetic using inverse modulo
- Precomputation of factorials and DP values
"""

m = 998244353
f = [1] * (300001)  # f[i] stores the number of almost perfect permutations of length i
fac = [1] * (300001)  # fac[i] stores i! mod m
fac[2] = 2
f[1] = 1
f[2] = 2
# Precompute f[i] using recurrence relation f[i] = f[i-1] + (i-1)*f[i-2]
for i in range(3, 300001):
    f[i] = (f[i - 1] + (i - 1) * f[i - 2]) % m
    fac[i] = (i * fac[i - 1]) % m


def calc(n):
    ans = 0
    k = 0
    # Iterate over all possible valid combinations of positions
    while 4 * k <= n:
        # Calculate contribution of current combination
        # Combine factorial terms with modular inverse
        ans += (fac[n - 2 * k] * pow(fac[n - 4 * k] * fac[k], -1, m) * f[n - 4 * k]) % m
        k += 1
    return ans % m


for _ in range(int(input())):
    n = int(input())
    print(calc(n))


# https://github.com/VaHiX/CodeForces/

File: CPP/ByRound/1726/1726_E_Almost_Perfect.cpp
// Problem: CF 1726 E - Almost Perfect
// https://codeforces.com/contest/1726/problem/E

/*
 * Problem: Count the number of "almost perfect" permutations of length n.
 * 
 * An almost perfect permutation satisfies: for all i, |p[i] - p^{-1}[i]| <= 1.
 * This means each element is at most one position away from its value in the inverse permutation.
 * 
 * Algorithm:
 * - Precompute factorials and inverse factorials for fast modular arithmetic.
 * - Use dynamic programming to compute base count of almost perfect permutations.
 * - Then, add corrections for cases involving cycles of length 4 that were overcounted.
 * 
 * Time Complexity: O(n) preprocessing + O(n) per query
 * Space Complexity: O(n)
 */

#include <cstdio>
#define P 998244353
int i, T, n, f[300005], I[300005], C, ans;

int main() {
  // Precompute factorials and inverse factorials
  f[0] = f[1] = I[1] = 1;
  for (i = 2; i <= 300000; i++)
    f[i] = (f[i - 1] + 1ll * (i - 1) * f[i - 2]) % P,  // DP recurrence for f[i]
    I[i] = 1ll * (P - P / i) * I[P % i] % P;  // Efficiently compute modular inverse

  scanf("%d", &T);
  while (T--) {
    scanf("%d", &n), C = 1, ans = f[n];  // Initialize C = 1 and ans = base count
    
    // Iterate over number of 4-cycles that can be formed
    for (i = 1; i <= n >> 2; i++)
      C = 1ll * C * (n - (i << 2) + 4) % P * (n - (i << 2) + 3) % P *  // Multiply by combinations of 4 elements
          (n - (i << 2) + 2) % P * (n - (i << 2) + 1) % P *  // These are combinations of 4 elements from remaining
          I[n - (i << 1) + 1] % P * I[n - (i << 1) + 2] % P * I[i] % P,  // Apply modular inverse to count combinations
      ans = (ans + 1ll * C * f[n - (i << 2)]) % P;  // Add corrected term to total count
    printf("%d\n", ans);
  }
}


// https://github.com/VaHiX/CodeForces/